pkg/code-generator/cmd/main.go

```diff
@@ -165,61 +156,96 @@ func collectProjectsFromRoot(root string, skipDirs []string) ([]model.ProjectCon
+	if err != nil {
+		return errors.Wrapf(err, "reading descriptors")
+	}
+
+addFiles:
+	for _, f := range desc.File {
+		// don't add the same proto twice, this avoids the issue where a dependency is imported multiple times
+		// with different import paths
+		for _, existing := range *descriptors {
+			if existing.GetName() == f.GetName() {
+				continue
+			}
+			existingCopy := proto.Clone(existing).(*descriptor.FileDescriptorProto)
+			existingCopy.Name = f.Name
+			if proto.Equal(existingCopy, f) {
+				continue addFiles
+			}
+		}
+		*descriptors = append(*descriptors, f)
+	}
+
+	return nil
+}
+
+func collectDescriptorsFromRoot(root string, customImports, skipDirs []string, wantCompile func(string) bool) ([]*descriptor.FileDescriptorProto, error) {
+	var descriptors []*descriptor.FileDescriptorProto
 
-	if err := filepath.Walk(root, func(protoFile string, info os.FileInfo, err error) error {
+	wg := sync.WaitGroup{}
+	errs := make(chan error)
+	err := filepath.Walk(root, func(protoFile string, info os.FileInfo, err error) error {
 		if !strings.HasSuffix(protoFile, ".proto") {
 			return nil
 		}
 		for _, skip := range skipDirs {
-			skipRoot, err := filepath.Abs(skip)
-			if err != nil {
-				return err
-			}
+			skipRoot := filepath.Join(root, skip)
 			if strings.HasPrefix(protoFile, skipRoot) {
-				log.Warnf("skipping detected proto %v", protoFile)
+				log.Warnf("skipping proto %v because it is %v is a skipped directory", protoFile, skipRoot)
 				return nil
 			}
 		}
-		imports, err := importsForProtoFile(root, protoFile, customImports)
-		if err != nil {
-			return err
-		}
-		imports = stringutils.Unique(imports)
-
-		// don't generate protos for non-project files
-		compile := wantCompile(protoFile)
-
-		if err := writeDescriptors(protoFile, tmpFile, imports, compile); err != nil {
-			return err
-		}
-		desc, err := readDescriptors(tmpFile)
-		if err != nil {
-			return err
-		}
-	addFiles:
-		for _, f := range desc.File {
-			// don't add the same proto twice, this avoids the issue where a dependency is imported multiple times
-			// with different import paths
-			for _, existing := range descriptors {
-				if existing.GetName() == f.GetName() {
-					continue
-				}
-				existingCopy := proto.Clone(existing).(*descriptor.FileDescriptorProto)
-				existingCopy.Name = f.Name
-				if proto.Equal(existingCopy, f) {
-					continue addFiles
-				}
+		wg.Add(1)
+		// parallelize parsing the descriptors as each one requires file i/o and is slow
+		go func() {
+			err := addDescriptorsForFile(&descriptors, root, protoFile, customImports, wantCompile)
+			wg.Done()
+			if err != nil {
+				log.Warnf("adding descriptors for file %v: %s", protoFile, err)
+				errs <- errors.Wrapf(err, "adding descriptors for file %v", protoFile)
 			}
-			descriptors = append(descriptors, f)
-		}
+		}()
 		return nil
-	}); err != nil {
+	})
+	if err != nil {
 		return nil, err
 	}
+	wg.Wait()
+	select {
+	case err := <-errs:
```

Comment:
there is a race here. wg.Done() is called before the error is pushed to the channel.
so it could be that Wait() will return while the error channel is still empty.

suggestion:
use a buffered channel: `errs := make(chan error, 1)`
use `defer wg.Done()` as the first line of the goroutine
instead of `wg.Wait()` do
```
go func() {
 wg.Wait()
 close(errs)
}()
```
when pushing the error use select to make sure the channel isnt full:
```
select {
case errs <- errors.Wrapf(err, "adding descriptors for file %v", protoFile):
default:
```
finally, then you can just receive from the channel
```
err := <- errs
```